backbone.js

```diff
@@ -1526,11 +1555,36 @@
     // treated as `null` to normalize cross-browser behavior.
     _extractParameters: function(route, fragment) {
       var params = route.exec(fragment).slice(1);
-      return _.map(params, function(param, i) {
+      var decodedParams = _.map(params, function(param, i) {
         // Don't decode the search params.
         if (i === params.length - 1) return param || null;
         return param ? decodeURIComponent(param) : null;
       });
+
+      if (this.useParamsObject) {
+        // get routeInfo, which has array of param names
+        var routeInfo = this._routeInfos[route.toString()];
+
+        var routeData = {
+          name: routeInfo.name
+        };
+
+        if (routeInfo.isRegExpRoute) {
+          // if original route is RegExp, everything goes into .params
+          routeData.params = decodedParams;
+        }
+        else {
+          routeData.params = _.object(routeInfo.paramNames, decodedParams);
+          var queryString = decodedParams[decodedParams.length - 1];
+          if (queryString) {
+            routeData.queryString = queryString;
+          }
+        }
+
+        return [routeData];
```

Comment:
So the handler would receive `(paramsObject, queryString)` as parameters, right?

We also discussed about this up here:
 https://github.com/jashkenas/backbone/pull/3815#discussion_r41050447